ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java

```diff
@@ -706,46 +766,132 @@ private void processNewPut(ScanResults results, IndexValue putValue) {
+                  EnumSet.of(PersistentIndex.IndexEntryType.TTL_UPDATE));
+          if (ttlUpdateValue != null) {
+            handleLogSegmentDeletedBucketUpdate(results, ttlUpdateValue, operationTimeInMs, operator * SUBTRACT);
+          }
+        } catch (StoreException e) {
+          logger.error(
+              "Failed to find TTL_UPDATE IndexValue for " + key + " when processing new DELETE: " + deleteValue, e);
+        }
       }
     }
   }
 
   /**
    * Helper function to process new UNDELETE entries and make appropriate updates to the given {@link ScanResults}.
    * @param results the {@link ScanResults} to apply the updates to
+   * @param key the {@link StoreKey} of the new UNDELETE
    * @param undeleteValue the {@link IndexValue} of the new UNDELETE
    * @param originalPutValue the {@link IndexValue} of the original PUT
+   * @param previousValue the {@link IndexValue} comes prior to the new UNDELETE
    */
-  private void processNewUndelete(ScanResults results, IndexValue undeleteValue, IndexValue originalPutValue) {
-    // TODO: future work to support online updates
+  private void processNewUndelete(ScanResults results, StoreKey key, IndexValue undeleteValue,
+      IndexValue originalPutValue, IndexValue previousValue) {
+    // First deal with container bucket
+    if (!previousValue.isDelete()) {
+      // This previous value is not DELETE, current UNDELETE would not change container bucket.
+      // There are several possibility when the previous value is not DELETE.
+      // 1. PUT UNDELETE: UNDELETE doesn't change any bucket
+      // 2. PUT TTL_UPDATE UNDELETE: UNDELETE doesn't change any bucket
+      // 3. PUT DELETE UNDELETE UNDELETE: the last UNDELETE doesn't change any bucket
+      // UNDELETE would resurrect a deleted IndexValue, if the previous value is not DELETE, then the IndexValue doesn't
+      // need to be resurrected.
+    } else {
+      handleContainerBucketUpdate(results, originalPutValue, undeleteValue.getOperationTimeInMs(), ADD);
+      handleContainerBucketUpdate(results, originalPutValue, originalPutValue.getExpiresAtMs(), SUBTRACT);
+    }
+
+    // Now deal with log segment bucket
+    // Current UNDELETE is always valid
+    results.updateLogSegmentBaseBucket(undeleteValue.getOffset().getName(), undeleteValue.getSize());
+    if (previousValue.isDelete()) {
+      // Previous IndexValue is a DELETE, it should already
+      // 1. Added it's size to the first bucket in the scan result
+      // 2. Subtracted original PUT's size from the previousValue.getOperationTimeInMs() bucket.
+      // We need to recover from these two operations, so
+      // 1. Subtract previous DELETE's size from first bucket in the scan result.
+      // 2. Add original PUT's size to the previousValue.getOperationTimeInMs() bucket.
+      processDeleteUpdateLogSegmentHelper(results, key, previousValue, originalPutValue, SUBTRACT);
```

Comment:
Let's consider a case: PUT, DELETE, UNDELETE and PUT expires after UNDELETE op time. Based on my understanding, it will go through following steps:
```
 Operation  | Base bucket | Delete Op bucket | Undelete Op bucket | Expired Bucket
    DELETE   | + D size        |  - P size                 |                                     |  + P size
 UNDELETE | - D + U size  |  + P size                |                                     |  + P size
```
Looks like the expired bucket adds PUT record size twice, did I misunderstand anything?

Reply:
if the current IndexValue is UNDELETE and the previous IndexValue is DELETE, then we will first revert the change made by DELETE, so for UNDELETE, at Expired Bucket, it will do -P size, not, +P size.